File: python/basic/sort_algorithm/quick_sort.py
# 快速排序
# 它的思想是取需要排序的数组中的第一个数作为基准，然后依次比较，把小于该基准数的都放左边，大于该基准数的都放右边。然后左右两边继续按照此方法进行，直到排序完成。

import random


def sort(lista, start, end): # 对列表中的某一段进行快排
    if start >= end: # 索引超出或相等时退出递归
        return
    low = start # 低位索引
    high = end # 高位索引
    mid = low # 基准数据的索引，初始值为地位索引
    while low < high:
        while low < high and lista[mid] > lista[low]: # 左边的数字比基准数小时将低位索引右移，直到低位索引所在的值比基准数大
            low += 1
        while low < high and lista[mid] < lista[high]: # 右边的数字比基准数大时将高位索引左移，直到高位索引所在的值比基准数小
            high -= 1
        lista[low], lista[high] = lista[high], lista[low] # 将高位索引的值与低位索引的值互换。这样可以继续进行比较。
    lista[mid], lista[low] = lista[low], lista[mid] # 上面的循环结束后需要把左边基准数据挪到低位索引，这样基准数据左边的数据都比基准数据小，右边的数据都比基准数据大。
    sort(lista, start, low-1) # 把基准数据左边的数据进行快排
    sort(lista, low+1, end) # 把基准数据右边的数据进行快排


def test(): # 测试函数
    a = list(range(120))
    for i in range(8):
        b = random.sample(a, 20)
        sort(b, 0, len(b)-1)
        print(b)


test()

File: golang/src/basic/algorithm/sort/QuickSort.go
package main

import "fmt"

func sort(nums []int, start, end int) {
	if start >= end {
		return
	}
	low := start + 1
	high := end
	mid := start
	for low < high {
		for low < high && nums[low] < nums[mid] {
			low++
		}
		for low < high && nums[high] > nums[mid] {
			high--
		}
		// low指针位置的数字不比mid位置小，high位置不比mid位置大。low和high换一个位置，可能它们还可以继续比
		nums[low], nums[high] = nums[high], nums[low]
	}
	// 此时low左边的数字都比mid位置的小，右边位置的数字都比mid位置的大
	// 8 1 2 3 4 5 6 10 12 13
	if nums[low] > nums[mid] {
		low--
	}
	// 与低位的数字进行位置调换
	nums[mid], nums[low] = nums[low], nums[mid]
	sort(nums, start, low-1)
	sort(nums, low+1, end)
}

func quickSort(nums []int) []int {
	sort(nums, 0, len(nums)-1)
	return nums
}

func main() {

	fmt.Println(quickSort([]int{10, 9, 8, 7, 6, 5}))
	fmt.Println(quickSort([]int{10, 19, 28, 37, 46, 55}))
	fmt.Println(quickSort([]int{8, 10, 22, 55, 32, 7, 3}))

}
